var/spack/repos/builtin/packages/m4/package.py

```diff
@@ -61,3 +66,24 @@ def configure_args(self):
             args.append('ac_cv_type_struct_sched_param=yes')
 
         return args
+
+    def test(self):
+        m4 = which('m4')
+        assert m4 is not None
+
+        tty.msg('test: Ensuring use of the installed executable')
+        m4_dir = os.path.dirname(m4.path)
+        assert m4_dir == self.prefix.bin
+
+        tty.msg('test: Checking version')
+        output = m4('--version', output=str.split, error=str.split)
+        version_regex = re.compile(r'm4(.+){0}'.format(self.spec.version))
+        assert version_regex.search(output)
+
+        tty.msg('test: Ensuring m4 runs')
+        currdir = os.getcwd()
+        hello_file = os.path.join(currdir, 'data', 'hello.m4')
+        output = m4(hello_file, output=str.split, error=str.split)
+        expected_file = os.path.join(currdir, 'data', 'hello.out')
+        with open(expected_file) as fd:
+            assert output == fd.read()
```

Comment:
These are four different tests in the same function. Was it considered to have an API based on decorators that can register tests? Something similar to:
```python
@test('Ensure use of installed executable')
def test_use(self):
    m4 = which('m4')
    m4_dir = os.path.dirname(m4.path)
    msg = 'the path of m4 is "{0}" while "{1}" was expected'
    assert m4_dir == self.prefix.bin￼, msg.format(m4_dir, self.prefix.bin)
```

Reply:
Developing further this idea (take this as a sort of brainstorming). It seems to me that the tests introduced by this PR are called after something has been installed, and as such they don't enter in the flow of concretization or installation.

This means that it is not strictly necessary to add further methods to `PackageBase` to handle tests but we could instead have a separate `PackageTester` class or some other entity  that is responsible for that. This opens the possibility to have tests in functions rather than in methods:
```python
def test_use(pkg):
    m4 = which('m4')
    m4_dir = os.path.dirname(m4.path)
    msg = 'the path of m4 is "{0}" while "{1}" was expected'
    assert m4_dir == self.prefix.bin￼, msg.format(m4_dir, self.prefix.bin
```
With a design like that we are more free to decide where to put tests. For instance they can reside in `package.py` if they are concise or they can be put in a separate file if they have a lot of details (like the `openmpi` case in this PR).

Reply:
> With a design like that we are more free to decide where to put tests. For instance they can reside in `package.py` if they are concise or they can be put in a separate file if they have a lot of details (like the `openmpi` case in this PR).

Since the goal is to have package maintainers maintain the install/smoke tests, there are advantages to having the functionality in `package.py` in that having them located in that module serves as a reminder to update them.  However, this rationale may be more relevant to packages with simple builds.